### PR TITLE
Updates to the new build system

### PR DESCRIPTION
**Pull request by ThadHouse**
Jenkins will fail until I update the scripts there, which I will do in a bit.

### CODE REVIEW

File: build.gradle
Comment (JLLeitschuh) [inline build.gradle:55]: Doesn't this cause this to get resolved eagerly at configuration time? Don't you want this to get resolved at task runtime?

File: config.gradle
Comment (JLLeitschuh) [inline config.gradle]: Why does this closure take an arg but never use it?

File: publish.gradle
Comment (JLLeitschuh) [inline publish.gradle]: Could just use `file("$rootDir/license.txt")`

File: publish.gradle
Comment (JLLeitschuh) [inline publish.gradle]: You might also want to check that the type is a `CopySpec`?

File: publish.gradle
Comment (JLLeitschuh) [inline publish.gradle]: Maybe turn this into a more strongly typed groovy function instead of using a closure since there are so many inputs? It would help others read it.

File: src/test/java/TestClass.java
Comment (JLLeitschuh) [inline src/test/java/TestClass.java]: What is this testing? Perhaps a bit more documentation if you are actually expecting something here.

File: src/test/java/TestClass.java
Comment (JLLeitschuh) [inline src/test/java/TestClass.java]: This seems to be in the default package. This is not really done even in test systems. You should put this test in its own package.

File: build.gradle
Comment (ThadHouse) [reply build.gradle:55]: It does not. The native builds handle eagerly configuring this. It does get resolved at runtime.

File: config.gradle
Comment (ThadHouse) [reply config.gradle]: I always forget in gradle you can ignore closure arguments. There are other cases where that config is used, but I can remove it from here.

File: publish.gradle
Comment (ThadHouse) [reply publish.gradle]: Good idea. Something similar to this is used in most of the repo's, so thats a good idea.

File: src/test/java/TestClass.java
Comment (ThadHouse) [reply src/test/java/TestClass.java]: Ok I will move it.

File: src/test/java/TestClass.java
Comment (ThadHouse) [reply src/test/java/TestClass.java]: This was more to verify that tests were actually running, and the test below is to make sure the JNI linking actually works. I'll remove the general one, and move the JNI one to it's own package.

File: publish.gradle
Comment (ThadHouse) [reply publish.gradle]: Yeah probably.

Comment (ThadHouse) [reply publish.gradle]: Actually the task it is looking at is not a copy task, so that doesn't work. This is one of the few cases where it is just hard to get something out of the model.

File: publish.gradle
Comment (ThadHouse) [reply publish.gradle]: I'm going to pass on this, because I use a few script globals, and switching to a function breaks that, with the workarounds looking super hacky

File: src/main/java/edu/wpi/first/wpilibj/networktables/NetworkTablesJNI.java
Comment (PeterJohnson) [inline src/main/java/edu/wpi/first/wpilibj/networktables/NetworkTablesJNI.java]: This seems like an unintentional revert.  Are you synced up to the latest master?

File: src/main/java/edu/wpi/first/wpilibj/networktables/NetworkTablesJNI.java
Comment (PeterJohnson) [inline src/main/java/edu/wpi/first/wpilibj/networktables/NetworkTablesJNI.java]: Ditto

**Review by PeterJohnson — CHANGES_REQUESTED** [review]

File: publish.gradle
Comment (PeterJohnson) [inline publish.gradle]: s/wpiutil/ntcore/

**Review by PeterJohnson — CHANGES_REQUESTED** [review]

File: build.gradle
Comment (AustinShalit) [inline build.gradle]: Should we bump to `4.0.2`?

Comment (JLLeitschuh) [reply build.gradle]: Yes

File: publish.gradle
Comment (JLLeitschuh) [reply publish.gradle]: What type is it then?

Comment (ThadHouse) [reply publish.gradle]: https://github.com/wpilibsuite/native-utils/blob/master/src/main/groovy/edu/wpi/first/nativeutils/rules/JNIConfigRules.groovy#L40
It's just a Task, it doesn't inherit from anything.

File: publish.gradle
Comment (ThadHouse) [reply publish.gradle]: Fixed

File: build.gradle
Comment (ThadHouse) [reply build.gradle]: fixed

File: publish.gradle
Comment (JLLeitschuh) [reply publish.gradle]: Reading through that code, just as an aside, its much better practice to create types for your tasks. Doing so allows you to use the annotations `@Input`, `@InputDirectory`, `@Output`, `@OutputDirectory` ect... That allow your build to do things better with task avoidance (speeding up the build).
Additionally, doing so would allow you to take advantage of the new gradle build cache where you can share the output of tasks on a central server (or even on your personal machine) to avoid running tasks that you've already run before.

Additionally, I worry that the lack of any sort of type system here will make this code unmaintainable by anyone in the future. Without a solid set of tests to prove that what the build is creating is correct it's much harder for new developers to dive in without the fear of breaking something.

Comment (ThadHouse) [reply publish.gradle]: How are the annotations different from just declaring Inputs and Outputs on the task, which I already do. Also, until all the native stuff supports the build cache (which none of it does currently), I don't think its necessary to take the extra work. I do agree that adding a custom type for this might be a good idea however. I just wish the team had commented on this weeks ago when I asked for help and suggestions rather then waiting until we were wanting to merge it.

File: publish.gradle
Comment (JLLeitschuh) [inline publish.gradle]: When does this execute? It looks like it runs during task configuration time, not at task run time? Is that what you want?

Comment (ThadHouse) [reply publish.gradle]: This does execute at run time. Otherwise copy and zip tasks would be absolutely useless.

Comment (JLLeitschuh) [reply publish.gradle]: Can you explain how that's possible? It seems like the receiver of the call to `from` is `project` not the task. As you said above, the task is of type `DefaultTask` which doesn't have a `from` method on it.

If the receiver of the `from` call is `project` then it would be executed immediately.

What is the receiver of the `from` call?

Comment (ThadHouse) [reply publish.gradle]: The cppSourcesZip. GitHub is just splitting the lines wierdly.

Comment (JLLeitschuh) [reply publish.gradle]: Oooooh. That makes so much more sense.

File: publish.gradle
Comment (ThadHouse) [reply publish.gradle]: @JLLeitschuh I created custom types for my tasks. I didn't add anything to them, since the default task worked just fine, but it does let me do type checking.

**Review by PeterJohnson — APPROVED** [review]

**Pull request merged**
